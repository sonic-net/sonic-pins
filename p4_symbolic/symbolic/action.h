// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically evaluate actions and their bodies.
// An action is represented as a boolean symbolic z3 expression over
// unconstrained symbolic parameters corresponding to its actual P4 parameters.

#ifndef P4_SYMBOLIC_SYMBOLIC_ACTION_H_
#define P4_SYMBOLIC_SYMBOLIC_ACTION_H_

#include <string>
#include <unordered_map>
#include <vector>

#include "google/protobuf/repeated_field.h"
#include "gutil/status.h"
#include "p4_pdpi/ir.pb.h"
#include "p4_symbolic/ir/ir.pb.h"
#include "p4_symbolic/symbolic/symbolic.h"
#include "p4_symbolic/symbolic/values.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace action {

// Symbolically evaluates the given action on the given symbolic parameters.
// This produces a symbolic expression on the symbolic parameters that is
// semantically equivalent to the behavior of the action on its concrete
// parameters.
absl::Status EvaluateAction(const ir::Action &action,
                            const google::protobuf::RepeatedPtrField<
                                pdpi::IrActionInvocation::IrActionParam> &args,
                            SymbolicPerPacketState *state,
                            values::P4RuntimeTranslator *translator,
                            const z3::expr &guard);

// Internal functions used to Evaluate statements and expressions within an
// action body. These are internal functions not used beyond this header and its
// associated source file.

// The scope of this action: maps local variable names to their symbolic values.
struct ActionContext {
  std::string action_name;
  std::unordered_map<std::string, z3::expr> scope;
};

// Performs a switch case over support statement types and call the
// appropriate function.
absl::Status EvaluateStatement(const ir::Statement &statement,
                               SymbolicPerPacketState *state,
                               ActionContext *context, const z3::expr &guard);

// Constructs a symbolic expression for the assignment value, and either
// constrains it in an enclosing assignment expression, or stores it in
// the action scope.
absl::Status EvaluateAssignmentStatement(
    const ir::AssignmentStatement &assignment, SymbolicPerPacketState *state,
    ActionContext *context, const z3::expr &guard);

// Constructs a symbolic expression corresponding to this value, according
// to its type.
absl::StatusOr<z3::expr> EvaluateRValue(const ir::RValue &rvalue,
                                        const SymbolicPerPacketState &state,
                                        const ActionContext &context);

// Extract the field symbolic value from the symbolic state.
absl::StatusOr<z3::expr> EvaluateFieldValue(const ir::FieldValue &field_value,
                                            const SymbolicPerPacketState &state,
                                            const ActionContext &context);

// Parse and format literal values as symbolic expression.
absl::StatusOr<z3::expr> EvaluateHexStr(const ir::HexstrValue &hexstr);

absl::StatusOr<z3::expr> EvaluateBool(const ir::BoolValue &bool_value);

absl::StatusOr<z3::expr> EvaluateString(const ir::StringValue &string_value);

// Looks up the symbolic value of the variable in the action scope.
absl::StatusOr<z3::expr> EvaluateVariable(const ir::Variable &variable,
                                          const ActionContext &context);

// Evaluate expression by recursively evaluating operands and applying the
// symbolic version of the operator to them.
absl::StatusOr<z3::expr> EvaluateRExpression(
    const ir::RExpression &expr, const SymbolicPerPacketState &state,
    const ActionContext &context);

}  // namespace action
}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_ACTION_H_
