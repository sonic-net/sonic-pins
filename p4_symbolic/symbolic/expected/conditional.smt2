(ingress) $got_cloned$: false
(ingress) $got_recirculated$: false
(ingress) ethernet.$extracted$: false
(ingress) ethernet.$valid$: (ite true true false)
(ingress) ethernet.dst_addr: ethernet.dst_addr
(ingress) ethernet.ether_type: ethernet.ether_type
(ingress) ethernet.src_addr: ethernet.src_addr
(ingress) scalars.$extracted$: false
(ingress) scalars.$valid$: false
(ingress) standard_metadata.$extracted$: false
(ingress) standard_metadata.$valid$: false
(ingress) standard_metadata._padding: standard_metadata._padding
(ingress) standard_metadata.checksum_error: #b0
(ingress) standard_metadata.deq_qdepth: #b0000000000000000000
(ingress) standard_metadata.deq_timedelta: #x00000000
(ingress) standard_metadata.egress_global_timestamp: #x000000000000
(ingress) standard_metadata.egress_port: standard_metadata.egress_port
(ingress) standard_metadata.egress_rid: #x0000
(ingress) standard_metadata.egress_spec: #b000000000
(ingress) standard_metadata.enq_qdepth: #b0000000000000000000
(ingress) standard_metadata.enq_timestamp: #x00000000
(ingress) standard_metadata.ingress_global_timestamp: #x000000000000
(ingress) standard_metadata.ingress_port: standard_metadata.ingress_port
(ingress) standard_metadata.instance_type: #x00000000
(ingress) standard_metadata.mcast_grp: #x0000
(ingress) standard_metadata.packet_length: standard_metadata.packet_length
(ingress) standard_metadata.parser_error: #x00000000
(ingress) standard_metadata.priority: #b000

(parsed) $got_cloned$: false
(parsed) $got_recirculated$: false
(parsed) ethernet.$extracted$: (ite true true false)
(parsed) ethernet.$valid$: (ite true true false)
(parsed) ethernet.dst_addr: ethernet.dst_addr
(parsed) ethernet.ether_type: ethernet.ether_type
(parsed) ethernet.src_addr: ethernet.src_addr
(parsed) scalars.$extracted$: false
(parsed) scalars.$valid$: false
(parsed) standard_metadata.$extracted$: false
(parsed) standard_metadata.$valid$: false
(parsed) standard_metadata._padding: standard_metadata._padding
(parsed) standard_metadata.checksum_error: #b0
(parsed) standard_metadata.deq_qdepth: #b0000000000000000000
(parsed) standard_metadata.deq_timedelta: #x00000000
(parsed) standard_metadata.egress_global_timestamp: #x000000000000
(parsed) standard_metadata.egress_port: standard_metadata.egress_port
(parsed) standard_metadata.egress_rid: #x0000
(parsed) standard_metadata.egress_spec: #b000000000
(parsed) standard_metadata.enq_qdepth: #b0000000000000000000
(parsed) standard_metadata.enq_timestamp: #x00000000
(parsed) standard_metadata.ingress_global_timestamp: #x000000000000
(parsed) standard_metadata.ingress_port: standard_metadata.ingress_port
(parsed) standard_metadata.instance_type: #x00000000
(parsed) standard_metadata.mcast_grp: #x0000
(parsed) standard_metadata.packet_length: standard_metadata.packet_length
(parsed) standard_metadata.parser_error: (ite (and true (not true)) #x00000002 #x00000000)
(parsed) standard_metadata.priority: #b000

(egress) $got_cloned$: false
(egress) $got_recirculated$: false
(egress) ethernet.$extracted$: (ite true true false)
(egress) ethernet.$valid$: (ite true true false)
(egress) ethernet.dst_addr: (ite (and true true (= ethernet.dst_addr #x000000000001))
     #x000000000002
     ethernet.dst_addr)
(egress) ethernet.ether_type: ethernet.ether_type
(egress) ethernet.src_addr: ethernet.src_addr
(egress) scalars.$extracted$: false
(egress) scalars.$valid$: false
(egress) standard_metadata.$extracted$: false
(egress) standard_metadata.$valid$: false
(egress) standard_metadata._padding: standard_metadata._padding
(egress) standard_metadata.checksum_error: #b0
(egress) standard_metadata.deq_qdepth: #b0000000000000000000
(egress) standard_metadata.deq_timedelta: #x00000000
(egress) standard_metadata.egress_global_timestamp: #x000000000000
(egress) standard_metadata.egress_port: (let ((a!1 (and true (not (and true (= ethernet.dst_addr #x000000000001)))))
      (a!2 (and true (not (= standard_metadata.ingress_port (concat #x00 #b0)))))
      (a!3 (ite (and true (= standard_metadata.ingress_port (concat #x00 #b0)))
                #b111111111
                #b000000000)))
(let ((a!4 (ite (and true (and true (= ethernet.dst_addr #x000000000001)))
                #b000000001
                (ite a!1 #b111111111 (ite a!2 #b111111111 a!3)))))
  (ite (not (= a!4 #b111111111)) a!4 standard_metadata.egress_port)))
(egress) standard_metadata.egress_rid: #x0000
(egress) standard_metadata.egress_spec: (let ((a!1 (and true (not (and true (= ethernet.dst_addr #x000000000001)))))
      (a!2 (and true (not (= standard_metadata.ingress_port (concat #x00 #b0)))))
      (a!3 (ite (and true (= standard_metadata.ingress_port (concat #x00 #b0)))
                #b111111111
                #b000000000)))
  (ite (and true (and true (= ethernet.dst_addr #x000000000001)))
       #b000000001
       (ite a!1 #b111111111 (ite a!2 #b111111111 a!3))))
(egress) standard_metadata.enq_qdepth: #b0000000000000000000
(egress) standard_metadata.enq_timestamp: #x00000000
(egress) standard_metadata.ingress_global_timestamp: #x000000000000
(egress) standard_metadata.ingress_port: standard_metadata.ingress_port
(egress) standard_metadata.instance_type: #x00000000
(egress) standard_metadata.mcast_grp: #x0000
(egress) standard_metadata.packet_length: standard_metadata.packet_length
(egress) standard_metadata.parser_error: (ite (and true (not true)) #x00000002 #x00000000)
(egress) standard_metadata.priority: #b000

(solver constraints)
; 
(set-info :status unknown)
(declare-fun standard_metadata.ingress_port () (_ BitVec 9))
(declare-fun ethernet.dst_addr () (_ BitVec 48))
(assert
 (let (($x56 (= standard_metadata.ingress_port (_ bv1 9))))
 (or (or false (= standard_metadata.ingress_port (_ bv0 9))) $x56)))
(assert
 (let (($x12 (= standard_metadata.ingress_port (concat (_ bv0 8) (_ bv0 1)))))
 (let (($x24 (and true $x12)))
 (let (($x17 (and true (not $x12))))
 (let ((?x42 (ite (and true (not (and true (= ethernet.dst_addr (_ bv1 48))))) (_ bv511 9) (ite $x17 (_ bv511 9) (ite $x24 (_ bv511 9) (_ bv0 9))))))
 (let (($x37 (= ethernet.dst_addr (_ bv1 48))))
 (let (($x38 (and true $x37)))
 (let (($x39 (and true $x38)))
 (let ((?x47 (ite $x39 (_ bv1 9) ?x42)))
 (let (($x59 (or (or false (= ?x47 (_ bv0 9))) (= ?x47 (_ bv1 9)))))
 (let (($x49 (= ?x47 (_ bv511 9))))
 (or $x49 $x59))))))))))))
(check-sat)
