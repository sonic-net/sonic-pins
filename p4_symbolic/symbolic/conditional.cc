// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically evaluate P4 conditionals and their
// branches.

#include "p4_symbolic/symbolic/conditional.h"

#include "p4_symbolic/symbolic/action.h"
#include "p4_symbolic/symbolic/operators.h"
#include "p4_symbolic/symbolic/util.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace conditional {

absl::StatusOr<SymbolicTrace> EvaluateConditional(
    const Dataplane data_plane, const ir::Conditional &conditional,
    SymbolicPerPacketState *state, values::P4RuntimeTranslator *translator,
    const z3::expr &guard) {
  // Evaluate the condition.
  action::ActionContext fake_context = {conditional.name(), {}};
  ASSIGN_OR_RETURN(
      z3::expr condition,
      action::EvaluateRValue(conditional.condition(), *state, fake_context));
  ASSIGN_OR_RETURN(z3::expr negated_condition, operators::Not(condition));

  // Build new guards for each branch.
  ASSIGN_OR_RETURN(z3::expr if_guard, operators::And(guard, condition));
  ASSIGN_OR_RETURN(z3::expr else_guard,
                   operators::And(guard, negated_condition));

  // Evaluate both branches.
  ASSIGN_OR_RETURN(SymbolicTrace if_trace,
                   control::EvaluateControl(data_plane, conditional.if_branch(),
                                            state, translator, if_guard));
  ASSIGN_OR_RETURN(
      SymbolicTrace else_trace,
      control::EvaluateControl(data_plane, conditional.else_branch(), state,
                               translator, else_guard));

  // Now we have two traces that need merging.
  // We should merge in a way such that the value of a field in the trace is
  // the one from the if branch if the condition is true, and the else branch
  // otherwise.
  return util::MergeTracesOnCondition(condition, if_trace, else_trace);
}

}  // namespace conditional
}  // namespace symbolic
}  // namespace p4_symbolic
