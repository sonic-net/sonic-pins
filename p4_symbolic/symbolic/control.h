// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file is responsible for evaluating a control construct in a P4
// program flow. A control construct could be a table match or a conditional
// leading to a table match.
//
// The bmv2 format suggests it can also be a direct action call, but that is
// unsupported by our IR and this file, supporting it should be easy if the
// need arise.
//
// This file is responsible for evaluating that control on the current
// evaluation state (essentially an assignment of header fields to symbolic
// values).
//
// Evaluating the control (e.g. via table::EvaluateTable(...)) returns a suffix
// SymbolicTrace: a trace that contains new values to header fields, as well as,
// table matches, for the control that was evaluated, and all next controls
// evaluated after it. This file returns that suffix trace as is. It is the
// responsibility of the calling code to extend that suffix to cover previous
// controls.
//
// Note that the first call to this file will return a complete trace.
//
// Control evaluation functions (e.g. table::EvaluateTable(...)) are responsible
// for calling this file again on the next control constructs that should be
// evaluated after they are evaluated.
//
// Example:
// Table A is described in BMV2 JSON so that after it is matched, the next
// control is Table B, which has no next control.
// Our symbolic pipeline will call this file on table A, which calls
// EvaluateTable(A, ...), which will evaluate A, and then call this file again
// on table B, which then calls EvaluateTable(B, ...).
// EvaluateTable(B, ...) will return a suffix trace to this file, consisting of
// the header fields values after B is evaluated, as well as the symbolic match
// for B. Our file will return it to EvaluateTable(A, ...), which will add the
// symbolic match for table A to it, and then return it the very first call,
// resulting in a complete trace.
//
// Conditionals operate similarly, but instead of appending a symbolic match
// to the suffix trace, they call this file on two controls corresponding to
// the conditional's two branches, then merge the two suffixes using symbolic
// if-then-else on a symbolic condition equivalent to that conditional.

#ifndef P4_SYMBOLIC_SYMBOLIC_CONTROL_H_
#define P4_SYMBOLIC_SYMBOLIC_CONTROL_H_

#include <string>

#include "gutil/status.h"
#include "p4_symbolic/symbolic/symbolic.h"
#include "p4_symbolic/symbolic/values.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace control {

// Evaluate the passed control construct.
absl::StatusOr<SymbolicTrace> EvaluateControl(
    const Dataplane &data_plane, const std::string &control_name,
    SymbolicPerPacketState *state, values::P4RuntimeTranslator *translator,
    const z3::expr &guard);

}  // namespace control
}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_CONTROL_H_
