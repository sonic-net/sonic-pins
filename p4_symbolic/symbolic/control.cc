// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file is responsible for evaluating a control construct in a P4
// program flow. A control construct could be a table match or a conditional
// leading to a table match.
//
// Check the header file for a detailed explanation.

#include "p4_symbolic/symbolic/control.h"

#include <vector>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "p4_pdpi/ir.pb.h"
#include "p4_symbolic/symbolic/conditional.h"
#include "p4_symbolic/symbolic/table.h"

namespace p4_symbolic {
namespace symbolic {
namespace control {

absl::StatusOr<SymbolicTrace> EvaluateControl(
    const Dataplane &data_plane, const std::string &control_name,
    SymbolicPerPacketState *state, values::P4RuntimeTranslator *translator,
    const z3::expr &guard) {
  // Base case: we got to the end of the evaluation, no more controls!
  if (control_name.empty()) {
    return SymbolicTrace{{}, Z3Context().bool_val(false)};
  }

  // Find out what type of control we need to evaluate.
  if (data_plane.program.tables().count(control_name) == 1) {
    // Table: call EvaluateTable on table and its entries.
    const ir::Table &table = data_plane.program.tables().at(control_name);
    std::vector<pdpi::IrTableEntry> table_entries;
    if (data_plane.entries.count(control_name) == 1) {
      table_entries = data_plane.entries.at(control_name);
    }
    return table::EvaluateTable(data_plane, table, table_entries, state,
                                translator, guard);
  } else if (data_plane.program.conditionals().count(control_name) == 1) {
    // Conditional: let EvaluateConditional handle it.
    const ir::Conditional &conditional =
        data_plane.program.conditionals().at(control_name);
    return conditional::EvaluateConditional(data_plane, conditional, state,
                                            translator, guard);
  } else {
    // Something else: unsupported.
    return absl::UnimplementedError(
        absl::StrCat("Unsupported control \"", control_name,
                     "\" is neither a table nor a conditional"));
  }
}

}  // namespace control
}  // namespace symbolic
}  // namespace p4_symbolic
