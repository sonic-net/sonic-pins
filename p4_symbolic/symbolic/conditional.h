// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Contains functions used to symbolically evaluate P4 conditionals and their
// branches.

#ifndef P4_SYMBOLIC_SYMBOLIC_CONDITIONAL_H_
#define P4_SYMBOLIC_SYMBOLIC_CONDITIONAL_H_

#include "gutil/status.h"
#include "p4_symbolic/ir/ir.pb.h"
#include "p4_symbolic/symbolic/control.h"
#include "p4_symbolic/symbolic/symbolic.h"
#include "p4_symbolic/symbolic/values.h"
#include "z3++.h"

namespace p4_symbolic {
namespace symbolic {
namespace conditional {

absl::StatusOr<SymbolicTrace> EvaluateConditional(
    const Dataplane data_plane, const ir::Conditional &conditional,
    SymbolicPerPacketState *state, values::P4RuntimeTranslator *translator,
    const z3::expr &guard);

}  // namespace conditional
}  // namespace symbolic
}  // namespace p4_symbolic

#endif  // P4_SYMBOLIC_SYMBOLIC_CONDITIONAL_H_
