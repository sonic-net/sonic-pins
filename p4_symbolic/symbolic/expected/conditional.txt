================================================================================
Finding packet for table MyIngress.t_1 and row -1
================================================================================
ingress_port = #b000000000
egress_port = #b000000001
trace:
dropped = 0
got cloned = 0
MyIngress.t_1 => was matched on entry -1
MyIngress.t_2 => was not matched!
MyIngress.t_3 => was matched on entry 0

ingress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = false
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000000
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

parsed_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000000
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

egress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000002
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000001
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000001
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000000
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000
________________________________________________________________________________

================================================================================
Finding packet for table MyIngress.t_2 and row -1
================================================================================
ingress_port = #b000000001
egress_port = #b000000001
trace:
dropped = 0
got cloned = 0
MyIngress.t_1 => was not matched!
MyIngress.t_2 => was matched on entry -1
MyIngress.t_3 => was matched on entry 0

ingress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = false
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

parsed_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

egress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000002
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000001
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000001
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000
________________________________________________________________________________

================================================================================
Finding packet for table MyIngress.t_3 and row -1
================================================================================
Cannot find solution!
________________________________________________________________________________

================================================================================
Finding packet for table MyIngress.t_3 and row 0
================================================================================
ingress_port = #b000000001
egress_port = #b000000001
trace:
dropped = 0
got cloned = 0
MyIngress.t_1 => was not matched!
MyIngress.t_2 => was matched on entry -1
MyIngress.t_3 => was matched on entry 0

ingress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = false
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

parsed_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000001
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000000
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000000
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000

egress_headers:
$got_cloned$ = false
$got_recirculated$ = false
ethernet.$extracted$ = true
ethernet.$valid$ = true
ethernet.dst_addr = #x000000000002
ethernet.ether_type = #x0000
ethernet.src_addr = #x000000000000
scalars.$extracted$ = false
scalars.$valid$ = false
standard_metadata.$extracted$ = false
standard_metadata.$valid$ = false
standard_metadata._padding = #b000
standard_metadata.checksum_error = #b0
standard_metadata.deq_qdepth = #b0000000000000000000
standard_metadata.deq_timedelta = #x00000000
standard_metadata.egress_global_timestamp = #x000000000000
standard_metadata.egress_port = #b000000001
standard_metadata.egress_rid = #x0000
standard_metadata.egress_spec = #b000000001
standard_metadata.enq_qdepth = #b0000000000000000000
standard_metadata.enq_timestamp = #x00000000
standard_metadata.ingress_global_timestamp = #x000000000000
standard_metadata.ingress_port = #b000000001
standard_metadata.instance_type = #x00000000
standard_metadata.mcast_grp = #x0000
standard_metadata.packet_length = #x00000000
standard_metadata.parser_error = #x00000000
standard_metadata.priority = #b000
________________________________________________________________________________

